<h1 align="center"><b>Naive Bayes</b>
<hr>

##Instalación de librerias

1. Necesitamos instalar la libreria tidyverse
> Nos servirá porque brinda un conjunto de paquetes útiles para la manipulación de datos, tanto a nivel de operaciones con *dplyr* o el manejo dentro de una tabla

In [ ]:
install.packages('tidyverse')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library(tidyverse)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.2     ✔ purrr   0.3.4
✔ tibble  3.0.3     ✔ dplyr   1.0.0
✔ tidyr   1.1.0     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



2. Necesitamos instalar la libreria tidytext
> Nos servirá porque nos va a ayudar a tokenizar las palabras que se encuentren dentro del tweet, incluso a generar columnas por palabras

In [ ]:
install.packages('tidytext')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘SnowballC’, ‘hunspell’, ‘tokenizers’, ‘janeaustenr’




In [ ]:
library(tidytext)

3. Necesitamos instalar la libreria naivebayes
> Nos servirá porque brinda las funciones necesarias para poder aplicar el modelo *Naive Bayes* 

In [ ]:
install.packages('naivebayes')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library(naivebayes)

naivebayes 0.9.7 loaded



4. Necesitamos instalar la libreria caret
> Nos servirá para poder generar nuestra matriz de confusión con los datos trabajados

In [ ]:
install.packages('caret')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘numDeriv’, ‘SQUAREM’, ‘lava’, ‘prodlim’, ‘iterators’, ‘data.table’, ‘gower’, ‘ipred’, ‘timeDate’, ‘foreach’, ‘plyr’, ‘ModelMetrics’, ‘reshape2’, ‘recipes’, ‘pROC’




In [ ]:
library(caret)

Loading required package: lattice


Attaching package: ‘caret’


The following object is masked from ‘package:purrr’:

    lift




5. Necesitamos instalar la libreria e1071
> Funciona en conjunto con la libreria *caret*, en este caso, brindando soporte para la matriz de confusión

In [ ]:
install.packages('e1071')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library(e1071)

##Exploración y Limpieza de datos


6. Cargamos nuestro dataset al servidor que ejecute R
>Se le indica como fileEncoding el tipo *latin1* para que pueda soportar los caracteres propios del español

In [ ]:
tweets_df <- read.csv(file = 'tweets_mx.csv', stringsAsFactors = F, fileEncoding = "latin1")

In [ ]:
tweets_df

status_id    screen_name    
1    8.324215e+17 lopezobrador_  
2    8.331306e+17 lopezobrador_  
3    8.335013e+17 lopezobrador_  
4    8.338952e+17 lopezobrador_  
5    8.341838e+17 lopezobrador_  
6    8.345660e+17 lopezobrador_  
7    8.352975e+17 lopezobrador_  
8    8.359036e+17 lopezobrador_  
9    8.364175e+17 lopezobrador_  
10   8.367459e+17 lopezobrador_  
11   8.371286e+17 lopezobrador_  
12   8.375080e+17 lopezobrador_  
13   8.377851e+17 lopezobrador_  
14   8.378487e+17 lopezobrador_  
15   8.382306e+17 lopezobrador_  
16   8.384377e+17 lopezobrador_  
17   8.391611e+17 lopezobrador_  
18   8.394921e+17 lopezobrador_  
19   8.396262e+17 lopezobrador_  
20   8.399606e+17 lopezobrador_  
21   8.403855e+17 lopezobrador_  
22   8.407467e+17 lopezobrador_  
23   8.409970e+17 lopezobrador_  
24   8.412962e+17 lopezobrador_  
25   8.413536e+17 lopezobrador_  
26   8.414772e+17 lopezobrador_  
27   8.417064e+17 lopezobrador_  
28   8.421327e+17 lopezobrador_  
29   8.425398e+17 lopezobrador_  
30   8.425755e+17 lopezobrador_  
⋮    ⋮            ⋮              
1320 9.834751e+17 JuanCruzFranco 
1321 9.834751e+17 ilvex          
1322 9.834751e+17 Haremymasmx    
1323 9.834751e+17 cata5581299141 
1324 9.834751e+17 IsaacPalafox   
1325 9.834751e+17 monkyher       
1326 9.834751e+17 JesusAlbertoZS 
1327 9.834751e+17 _victoriacetina
1328 9.834751e+17 dibala01       
1329 9.834751e+17 OswaldoCruzCoro
1330 9.834751e+17 peabooddy      
1331 9.834751e+17 famoushorse    
1332 9.834751e+17 polodelara     
1333 9.834751e+17 DanielPensamien
1334 9.834751e+17 BecSantus      
1335 9.834751e+17 LizzyTorress   
1336 9.834751e+17 fer_hool       
1337 9.834751e+17 niehauscanary  
1338 9.834751e+17 PaolaVirrueta  
1339 9.834751e+17 HermanosVazquez
1340 9.834751e+17 davidtinoco1975
1341 9.834751e+17 GR_ABIELA      
1342 9.834751e+17 malamujerrp    
1343 9.834751e+17 StephanySexy2  
1344 9.834751e+17 HabeshaP       
1345 9.834751e+17 DafneFCh       
1346 9.834751e+17 09osuna        
1347 9.834751e+17 louder_thancat 
1348 9.834751e+17 lilyindescai   
1349 9.834751e+17 perecerda      
     text                                                                                                                                                                                                                                                                                         
1    Josefina Vázquez Mota debe informar qué hizo con los mil millones de pesos que le autorizó EPN. Basta de corrupción e hipocresía PRIANISTA.                                                                                                                                                  
2    Ya ni la burla perdonan: bajaron 2 centavos la gasolina. En Guatemala no hay petróleo y es más barata la gasolina que en México. Corruptos.                                                                                                                                                  
3    Martín Moreno no votará por mí. Comprendo. Es un mal escritor dedicado a explotar la ignorancia y el conservadurismo de la sociedad. Gracias                                                                                                                                                 
4    En Chicago dije que iremos a Nueva York (ONU) y a Washington (CIDH) el 14 y 15 de marzo a denunciar órdenes-Trump https://t.co/nTeKURAIzv                                                                                                                                                    
5    Entérate y apoya con tu firma la denuncia contra las órdenes de Donald Trump https://t.co/FEICSPQ4pn                                                                                                                                                                                         
6    Qué república ni qué ocho cuartos, es la monarquía de la moronga azul. Corrupta e hipócrita. https://t.co/jYqIi4dFL8                              

7. Usaremos gsub para que con el uso de expresiones regulares podamos quitar los links que se puedan encontrar dentro de los tweets

In [ ]:
quitar_url <- function(texto) {
    gsub("\\<http\\S*\\>|[0-9]", " ", texto)
}

##Pasando a matriz dispersa

8. Convertiremos nuestros datos en una matriz dispersa, por lo que tendremos que segmentar cada palabra en columnas, hacer cuenta de la aparición de cada palabra y se le dara un formato ancho

In [ ]:
tweets_df %>%
  unnest_tokens(input = "text", output = "palabra") %>%
  count(screen_name, status_id, palabra) %>%
  spread(key = palabra, value = n)

screen_name     status_id    _mx _paolimeow _siqure 00 03 04npkn59na
1    __dianapatricia 9.834750e+17 NA  NA         NA      NA NA NA        
2    _vichoo1        9.834750e+17 NA  NA         NA      NA NA NA        
3    _victoriacetina 9.834751e+17 NA  NA         NA      NA NA NA        
4    09osuna         9.834751e+17 NA  NA         NA      NA NA NA        
5    1javierespinoza 9.834750e+17 NA  NA         NA      NA NA NA        
6    Accountant_Job  9.834750e+17 NA  NA         NA      NA NA NA        
7    AcuarioJac      9.834750e+17 NA  NA         NA      NA NA NA        
8    agenciasanluis  9.834750e+17 NA  NA         NA      NA NA NA        
9    Alexvaldesp     9.834750e+17 NA  NA         NA      NA NA NA        
10   Allisonband     9.834750e+17 NA  NA         NA      NA NA NA        
11   amediumplace    9.834750e+17 NA  NA         NA      NA NA NA        
12   amediumplace    9.834750e+17 NA  NA         NA      NA NA NA        
13   AnahuacOaxaca   9.834750e+17 NA  NA         NA      NA NA NA        
14   AndreaOchoaM    9.834750e+17 NA  NA         NA      NA NA NA        
15   Angel_BM        9.834750e+17 NA  NA         NA      NA NA NA        
16   angelasaju      9.834750e+17 NA  NA         NA      NA NA NA        
17   angelmarielle_  9.834751e+17 NA  NA         NA      NA NA NA        
18   AntequeraDiario 9.834750e+17 NA  NA         NA      NA NA NA        
19   Arresnita       9.834750e+17 NA  NA         NA      NA NA NA        
20   azamar1         9.834750e+17 NA  NA         NA      NA NA NA        
21   Banking_job     9.834750e+17 NA  NA         NA      NA NA NA        
22   baxtersa23      9.834750e+17 NA  NA         NA      NA NA NA        
23   BecSantus       9.834751e+17 NA  NA         NA      NA NA NA        
24   BGoldenVIP      9.834750e+17 NA  NA         NA      NA NA NA        
25   Bietan_jarrai59 9.834750e+17 NA  NA         NA      NA NA NA        
26   BrendaBCG12     9.834750e+17 NA  NA         NA      NA NA NA        
27   BrendaBCG12     9.834751e+17 NA  NA         NA      NA NA NA        
28   casaverde4piso  9.834750e+17 NA  NA         NA      NA NA NA        
29   cata5581299141  9.834751e+17 NA  NA         NA      NA NA NA        
30   CEASMx          9.834750e+17 NA  NA         NA      NA NA NA        
⋮    ⋮               ⋮            ⋮   ⋮          ⋮       ⋮  ⋮  ⋮         
1319 UNAM_MX         9.831324e+17 NA  NA         NA      NA NA NA        
1320 UNAM_MX         9.831475e+17 NA  NA         NA      NA NA NA        
1321 UNAM_MX         9.831626e+17 NA  NA         NA      NA NA NA        
1322 UNAM_MX         9.831777e+17 NA  NA         NA      NA NA NA        
1323 UNAM_MX         9.831852e+17 NA  NA         NA      NA NA NA        
1324 UNAM_MX         9.833138e+17 NA  NA         NA      NA NA NA        
1325 UNAM_MX         9.833287e+17 NA  NA         NA      NA NA NA        
1326 UNAM_MX         9.833363e+17 NA  NA         NA      NA NA NA        
1327 UNAM_MX         9.833440e+17 NA  NA         NA      NA NA NA        
1328 UNAM_MX         9.833513e+17 NA  NA         NA      NA NA NA        
1329 UNAM_MX         9.833590e+17 NA  NA         NA      NA NA NA        
1330 UNAM_MX         9.833665e+17 NA  NA         NA      NA NA NA        
1331 UNAM_MX         9.833740e+17 NA  NA         NA      NA NA NA        
1332 UNAM_MX         9.833894e+17 NA  NA         NA      NA NA NA        
1333 UNAM_MX         9.834042e+17 NA  NA         NA      NA NA NA        
1334 UNAM_MX         9.834118e+17 NA  NA         NA      NA NA NA        
1335 UNAM_MX         9.834196e+17 NA  NA         NA      NA NA NA        
1336 UNAM_MX         9.834269e+17 NA  NA         NA      NA NA NA        
1337 UNAM_MX         9.834345e+17 NA  NA         NA      NA NA NA        
1338 UNAM_MX         9.834495e+17 NA  NA         NA      NA NA NA        
1339 UNAM_MX         9.834571e+17 NA  NA         NA      NA NA NA        
1340 UNAM_MX         9.834647e+17 NA  NA         NA      NA NA NA        
1341 UNAM

9. Al ser un proceso repetitivo lo hacemos función y aplicamos el filtro para quitar links y de paso eliminamos columnas innecesarias

In [ ]:
crear_matriz <- function(tabla) {
  tabla %>%
    mutate(text = quitar_url(text)) %>%
    unnest_tokens(input = "text", output = "palabra") %>%
    count(screen_name, status_id, palabra) %>%
    spread(key = palabra, value = n) %>%
    select(-status_id)
}

##Naive Bayes

10. Lo que se busca es poder predecir si un tweet le pertenece a un usuario en particular, solo se usarán 2 categorías, en caso de no pertenecer, simplemente se le pone "Otro"

In [ ]:
ejemplo_matriz <-
  tweets_df %>%
  mutate(screen_name = ifelse(screen_name == "MSFTMexico", screen_name, "Otro"),
         screen_name = as.factor(screen_name)) %>%
  crear_matriz

###Sets de entrenamiento y prueba

11.Se van a dividir los datos de modo que tengamos 70% de ellos en el set de entrenamiento y el resto en el set de prueba. 
>Se usa seed para que sea reproducible y lo colocamos antes de obtener el primer set

In [ ]:
set.seed(2001)
ejemplo_entrenamiento <- sample_frac(ejemplo_matriz, .7)
ejemplo_prueba <- setdiff(ejemplo_matriz, ejemplo_entrenamiento)

###Aplicación del modelo

12. Llamamos a la función naive_bayes y le pasamos la variable objetivo para clasificar y los datos que serán usados

In [ ]:
ejemplo_modelo <- naive_bayes(formula = screen_name ~ .,  data = ejemplo_entrenamiento)

###Haciendo predicciones

13.Para realizar predicciones usaremos la funcion *predict* y le pasamos el modelo y datos nuevos, por lo que le pasaremos el set de prueba

In [ ]:
ejemplo_prediccion <- predict(ejemplo_modelo, ejemplo_prueba)

Warning message:
“predict.naive_bayes(): more features in the newdata are provided as there are probability tables in the object. Calculation is performed based on features to be found in the tables.”


14. Para analizar el éxito del modelo, crearemos una matriz de confusión, le pasamos el vector con las predicciones y los valores reales de screen_name

In [ ]:
confusionMatrix(ejemplo_prediccion, ejemplo_prueba[["screen_name"]])

Confusion Matrix and Statistics

            Reference
Prediction   MSFTMexico Otro
  MSFTMexico         48   20
  Otro               11  309
                                          
               Accuracy : 0.9201          
                 95% CI : (0.8885, 0.9451)
    No Information Rate : 0.8479          
    P-Value [Acc > NIR] : 1.404e-05       
                                          
                  Kappa : 0.7084          
                                          
 Mcnemar's Test P-Value : 0.1508          
                                          
            Sensitivity : 0.8136          
            Specificity : 0.9392          
         Pos Pred Value : 0.7059          
         Neg Pred Value : 0.9656          
             Prevalence : 0.1521          
         Detection Rate : 0.1237          
   Detection Prevalence : 0.1753          
      Balanced Accuracy : 0.8764          
                                          
       'Positive' Class : MSFTMexico     

>Obtuvimos una precisión del 92.01%, por lo que no se realizó algún ajuste en el modelo. También se obtuvo 93.92% de Especificidad, que corresponde en este caso a la categoria "Otro"